lib/dotcom/schedule_finder.ex

```diff
@@ -280,4 +285,179 @@ defmodule Dotcom.ScheduleFinder do
+
+    formatted_date =
+      single_date
+      |> Date.from_iso8601!()
+      |> format_tiny_date()
+
+    %{
+      service
+      | description: "#{date_note}, #{formatted_date}"
+    }
+  end
+
+  defp add_single_date_description(service), do: service
+
+  defp format_tiny_date(date), do: format!(date, :month_day_short)
+
+  defp adjust_planned_description(%{typicality: :planned_disruption} = service) do
+    dates =
+      if service.start_date == service.end_date do
+        " (#{format_tiny_date(service.start_date)})"
+      else
+        " (#{format_tiny_date(service.start_date)} through #{format_tiny_date(service.end_date)})"
+      end
+
+    Map.update!(service, :description, &(&1 <> dates))
+  end
+
+  defp adjust_planned_description(service), do: service
+
+  defp tag_next_available(service_descriptions, current_date) do
+    if Enum.any?(service_descriptions, &(&1.is_now? || &1.next_available?)) do
+      service_descriptions
+    else
+      next_date = Date.shift(current_date, day: 1)
+
+      service_descriptions
+      |> Enum.map(&Map.put(&1, :next_available?, next_date in &1.dates))
+      |> tag_next_available(next_date)
+    end
+  end
+
+  defp sort_services_by_date(service, start_date) do
+    {ordered_names(service.name), Date.to_string(start_date)}
+  end
+
+  defp ordered_names(name) do
+    [
+      "Weekday",
+      "Monday - Thursday",
+      "Friday",
+      "Saturday",
+      "Sunday",
+      "Holiday"
+    ]
+    |> Enum.find_index(&String.contains?(name, &1))
+  end
+
+  defp service_group(%ServiceDescription{
+         service: %Service{typicality: :holiday_service}
+       }),
+       do: "Holiday Schedules"
+
+  defp service_group(%ServiceDescription{
+         service: %Service{typicality: :extra_service}
+       }),
+       do: "Extra Service"
+
+  defp service_group(%ServiceDescription{
+         service: %Service{typicality: :planned_disruption} = service
+       }) do
+    "Planned Work #{service.rating_description}"
+  end
+
+  defp service_group(%ServiceDescription{
+         service: service
+       }) do
+    today = Dotcom.Utils.ServiceDateTime.service_date()
+
+    if Service.in_current_rating?(service, today) do
+      ends_or_starts =
+        if service.rating_end_date do
+          "ends #{format_tiny_date(service.rating_end_date)}"
+        else
+          "starts #{format_tiny_date(service.rating_start_date)}"
+        end
+
+      "#{service.rating_description} Schedules, #{ends_or_starts}"
+    else
+      if Service.in_future_rating?(service, today) do
+        "#{service.rating_description} Schedules, starts #{format_tiny_date(service.rating_start_date)}"
+      else
+        "Other Schedules"
+      end
+    end
+  end
+
+  defp service_group(_), do: "Other Schedules"
+
+  defp sort_groups({group_name, _}) do
+    [
+      "Schedules, ends",
+      "Schedules, starts",
+      "Holiday",
+      "Extra"
+    ]
+    |> Enum.find_index(&String.contains?(group_name, &1))
+  end
```

Comment:
At a glance, this looks like the sort of thing [that bit us](https://github.com/mbta/dotcom/pull/2716#discussion_r2326064661) with the `Now: ` prefix in system status. Specifically, I think this will preclude translating `Schedules` and `starts`/`ends`, since once those strings are translated, `String.contains?` won't work anymore.

I think the solution is to push string-logic to the very last possible moment before they land in an HTML template, so instead of constructing and then parsing strings in the backend, I would try to have this code construct more elaborate structs, and serialize them as strings in the frontend.

Reply:
You're totally right!